### ISSUE TITLE
Support steady state runs

### ISSUE THREAD

**Issue by gijsber** [Deltares/Ribasim#150]

**Comment by visr** [comment]
This can be interesting for simplified usage with large timesteps where changes in storage don't matter, except for large reservoirs.

**Comment by visr** [comment]
I'm thinking about a configuration option like this:

```toml
[solver]
steady_state = true     # optional, default false
```

When `starttime == endtime` this would just do a steady state run. When `starttime > endtime` it will calculate the steady state using `Basin / state` as `u0`, and then use the steady state as `u0` for the transient simulation. If it is very robust to initial guesses perhaps we can even make `Basin / state` optional, but for now I would just use it.

There is some documentation here:

https://docs.sciml.ai/DiffEqDocs/stable/solvers/steady_state_solve/
https://docs.sciml.ai/DiffEqDocs/stable/types/steady_state_types/
https://docs.sciml.ai/NonlinearSolve/stable/

**Comment by visr** [comment]
Yeah `steady_state="first"` or `steady_state="all"`. Is probably better. This issue should only implement `first`, but at least then we have an obvious way of specifying repeated steady states. In that scenario `dt` would probably be required, such that we know the times at which to solve the steady states.

**Comment by Huite** [comment]
> A steady state might not always exist

This is correct: if you specify an inflow but no outflows, there is no steady-state. But this is almost certainly the sign of a wrongly schematized model. The transient case will run, but will produce bad results. Having steady-states blow up is actually a way of detecting such faults!

> For certain models there might only be a steady state when certain basins are empty, which might not be a desirable steady state.

Empty is fine.

The "absoluteness" of steady-states is exactly what makes them an excellent tool to understand your model setup.

**Comment by SouthEndMusic** [comment]
This code looks for a steady state (positive storage enforced):

```julia
model = Ribasim.Model(toml_path)
function steady_initial_state!(du, u, p, _)
    Ribasim.water_balance!(du, u, p, 0.0)
    du .^= 2
    du.storage += max.(-du.storage, 0) .^ 2
end

prob = NonlinearProblem{true}(steady_initial_state!, model.integrator.u, model.integrator.p)
sol = solve(prob, NewtonRaphson(linesearch=HagerZhang()))
display(sol)
du = zero(model.integrator.u)
Ribasim.water_balance!(du, sol.u, model.integrator.p, 0.0)
du
```

It doesn't converge with default tolerances but maybe it is good enough for our purposes. What are sensible tolerances?

**Comment by SouthEndMusic** [comment]
The solver algorithms for finding the steady state feel too much like a black box to me, especially for large models. Therefore I tried to come up with easy to understand necessary conditions for the existence of a steady state we can check before trying to compute it. This can also give nice feedback to the modeler.

## Idea 1: test steady state feasibility

We talked about 'checking whether the water can leave the model', and in that direction I came up with the following idea: we can set up a JuMP problem for the flow through whole model with 'feasibility' objective, with:

- An equality constraint per basin for inflow = outflow, taking into account forcings and flow boundaries;
- Allowing unlimited outflow into terminals and level boundaries

Of course this has its own subtle problems and takes some time to implement, let me know what you think.

## Idea 2: constrained root finding

Finding the steady state with a root-finding algorithm on the RHS has its own problems: getting stuck in local minima, converging to infeasible solutions etc. To give this method the best chance of success, we can enforce certain properties of the solution. For instance enforcing non-negative storages is easy to do (see latest commit in PR). Let me know what properties of the solution you think we can enforce based on the model schematization.

## Idea 3: custom breakoff conditions

Steady states can also be computed by fixing the transient parameters and letting the model converge to a steady state by simulating. This is (mostly) what `SteadyStateDiffEq.DynamicSS` does under the hood, with a dedicated callback for checking convergence. We could do something similar, but make our own callbacks which test (somehow) whether it's likely/unlikely whether convergence to a steady state is actually happening/going to happen.

**Comment by visr** [comment]
Idea 2 sounds the best to me. Besides non-negative storages, I cannot think of other constraints right now. Is that one not enough? Either S goes to 0 or Inf. And Inf means there is no steady state. But probably I'm just not aware of the subtleties. What do the local minima and infeasible solutions look like? Have you see local minima with Inf storage, where you know there is an actual steady state?

**Comment by Huite** [comment]
I have the feeling the scope is getting widened? We just want a steady-state. If the model is poorly schematized such that a steady-state cannot be found, it will not converge. Stating that in clear terms such that anyone will understand is a separate issue.

Idea 1 seems rather complicated, and requires building a kind of "shadow model".

 > Finding the steady state with a root-finding algorithm on the RHS has its own problems: getting stuck in local minima, converging to infeasible solutions etc.

This is a general problem that applies both to steady and transient runs, right? Not just for steady-state.

> The solver algorithms for finding the steady state feel too much like a black box to me, especially for large models.

Why? If we're simplifying to just ImplicitEuler, the concept in principle is straightforward (find u such du/dt = 0, which is just rootfinding like Newton-Rhapson does). Mathematically the problem is harder to solve than transient flow though, which is why you might need tricks like "pseudo-transient-continuation" (PTC) where you add a little bit of storage change to get to your next iteration.

**Comment by SouthEndMusic** [comment]
I haven't done many very specific experiments, but:
- I don't trust the root finding approach to always find the steady state if it exists
- The simulation approach is most of a black box, because that function internally creates an ODE problem and solves it (with a special callback for detecting convergence) without any feedback.

You say that a steady state should exist if water can sufficiently leave the model, and I tend to agree because oscillations preventing convergence would require higher order derivatives to be present in the ODE (I think). PID control can mess that up but for that we can fix the integral value to 0.

Maybe it's best if we have a more relaxed convergence criterion in the simulation case: don't look for du = 0, just look for the sum of the storages converging (I.e. the sum of du.storage converging to 0).

**Comment by Huite** [comment]
I just discussed this with @SouthEndMusic:

What wasn't clear to me, but was clarified:

* DynamicSS is just a sequence of timesteps until du/dt approaches 0. This works across solvers; with an explicit method you just keep advancing.
* Rootfinder solves a different problem and will solve for du/dt = 0.

What I hadn't considered properly yet is the interaction with various callbacks. There's no need to update the boundary condition values during the initial steady-state, but in my thinking DiscreteControl **should** be involved somehow. Bart mentioned that initially DiscreteControl is evaluated initially, but not changed. I don't think this is the desired behavior:

* Let's assume we have some stage-controlled area like a polder
* If in the initial condition the polder is dry, the pump Q = 0.
* If water is draining from another area, this means that there's no steady-state possible as the water can only accumulate.
* Alternatively, if the initial condition is wet, the pump Q = capacity.
* This will result in a steady-state with near dry conditions, as the reduction factor balances inflow with outflow.

To note: I think we (as hydrologists) generally consider that there's a steady-state available given a physical (!) system with boundary conditions. This generally results in a single steady-state, but I would not be surprised if complex control measures can to lead to non-unique/equifinal steady-states. This makes things more difficult conceptually, since path dependence either via initial condition or through different solver setttings may result in a different steady-state.

Conclusion 1: discrete control will almost always(?) be unsuitable for steady-state computation. The essential idea with discrete control is that e.g. a pump switches on or off; and through doing so, it ends up varying around some set point level.
In the hydrological context, we could call that setpoint the steady-state level (something reflecting an equilibrium) but it requires a pump that is on 30%-off 70%, which cannot be reflected in a DiscreteControl way.

Conclusion 2: Taking 1 into account, a "hydrological steady-state" is dependent on control measures. This means that unlike for a transient simulation, to solve it properly in a Newton-Rhapson manner we should incorporate control terms in the Jacobian. This does not seem attractive to me. A "pseudo-transient" approach (like DynamicSS) seems much easier to me; the control callback would be called at every pseudo-timestep.

In terms of solutions:

* We have to accept that there are multiple steady-state solutions possible for a given model. I don't think this is what hydrologists are used to seeing: they mostly deal with natural/physical systems rather than control systems. So it probably requires good documentation, ideally even an example where two different initial states result in two separate solutions.
* I don't have good ideas for DiscreteControl. The only thing I can come up with is to make DiscreteControl continuous for steady-state. On/Off can be turned into a piecewise linear relation, but it's not trivial deciding on a slope. For a pump, you could argue that Qcapacity = volume at set point, and Q0 = volume when empty, but likely not every control measure has straightforward or intuitive points like that to choose.
